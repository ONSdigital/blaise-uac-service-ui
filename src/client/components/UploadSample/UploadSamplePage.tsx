import React, {ReactElement, useState} from "react";
import {Formik, Form} from "formik";
import {Link, Route, Switch} from "react-router-dom";
import App from "../../../App";
import SelectFile from "./Sections/SelectFile";
import InstrumentName from "./Sections/InstrumentName";
import {ONSPanel, ONSButton, StyledFormErrorSummary} from "blaise-design-system-react-components";
import {uploadFile} from "../../upload-file";
import UploadFailed from "./Sections/UploadFailed";
import UploadSuccessful from "./Sections/UploadSuccessful";

function UploadSamplePage(): ReactElement {
    const [instrumentName, setInstrumentName] = useState<string>();
    const [file, setFile] = useState<File>();
    const [activeStep, setActiveStep] = useState(0);

    function _renderStepContent(step: number) {
        switch (step) {
            case 1:
                return (
                    <UploadSuccessful instrumentName={instrumentName}/>
                );
            case 2:
                return (
                    <UploadFailed instrumentName={instrumentName}/>
                );
        }
    }

    async function _handleSubmit() {
        const result = await uploadFile(instrumentName, file);
        if(result === true) {
            setActiveStep(1);
        }
        else {
            setActiveStep(2);
        }
    }

    return (
        <>
            {/*<Breadcrumbs BreadcrumbList={*/}
            {/*    [*/}
            {/*        {link: "/", title: "Home"},*/}
            {/*    ]*/}
            {/*}/>*/}

            <main id="main-content" className="page__main u-mt-no">
                <Switch>
                    <Route path="/app">
                        <App/>
                    </Route>
                    <Route path="/">

                        <ul className="list list--bare list--inline u-mt-m">
                            <li className="list__item">
                                <Link to="/app" id="deploy-questionnaire-link">
                                    Home
                                </Link>
                            </li>
                        </ul>
                    </Route>
                </Switch>

                <Formik
                    validateOnBlur={false}
                    validateOnChange={false}
                    initialValues={{override: ""}}
                    onSubmit={_handleSubmit}>

                    <Form id={"formID"}>
                        <div>
                            <h1 className="u-mt-s">
                                Select instrument sample
                            </h1>

                            <ONSPanel>
                                <p>
                                    When a sample file is selected and you continue to upload this sample file, <b>this
                                    may take a few minutes</b>.
                                    <br/>
                                    <br/>
                                    Given this, <b>do not navigate away</b> from this page during this process. You will
                                    be
                                    re-directed
                                    when there is an update regarding the deploy of the questionnaire.
                                </p>
                            </ONSPanel>
                        </div>
                        <div className="u-mt-m">
                            <StyledFormErrorSummary/>
                            <InstrumentName
                                instrumentName={instrumentName}
                                setInstrumentName={setInstrumentName}/>
                            <SelectFile
                                file={file}
                                setFile={setFile}
                                loading={false}/>
                        </div>
                        {_renderStepContent(activeStep)}
                        <div className="btn-group u-mt-m">
                            <ONSButton
                                id={"upload-sample-button"}
                                submit={true}
                                primary={true} label={"Continue"}/>
                        </div>
                    </Form>

                </Formik>
            </main>
        </>
    );
}

export default UploadSamplePage;
