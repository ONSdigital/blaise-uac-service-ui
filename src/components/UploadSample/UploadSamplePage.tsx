import React, {ReactElement, useState} from "react";
import {Formik, Form} from "formik";
import {Link, Route, Switch} from "react-router-dom";
import App from "../../App";
import SelectFile from "./Sections/SelectFile";
import InstrumentName from "./Sections/InstrumentName";
import {ONSButton} from "blaise-design-system-react-components";
import {sampleFileAlreadyExists, generateUacCodesForSampleFile} from "../../client/file-functions";
import UploadFailed from "./Sections/UploadFailed";
import FileExists from "./Sections/FileExists";
import DownloadUacFile from "./Sections/DownloadUacFile";
import {Datas} from "react-csv-downloader/dist/esm/lib/csv";

function UploadSamplePage(): ReactElement {
    const [instrumentName, setInstrumentName] = useState<string>();
    const [overwrite, setOverwrite] = useState<string>();
    const [file, setFile] = useState<File>();
    const [activeStep, setActiveStep] = useState(0);
    const [downloadData, setDownloadData] = useState<Datas>([]);


    function _renderStepContent(step: number) {
        switch (step) {
            case 0:
                return (<InstrumentName instrumentName={instrumentName} setInstrumentName={setInstrumentName}/>);
            case 1:
                return (
                    <FileExists instrumentName={instrumentName} overwrite={overwrite} setOverwrite={setOverwrite}/>);
            case 2:
                return (<SelectFile file={file} setFile={setFile}/>);
            case 3:
                return (<DownloadUacFile instrumentName={instrumentName} downloadData={downloadData}/>);
            case 4:
                return (<UploadFailed instrumentName={instrumentName}/>);
        }
    }

    async function generateUacCodesAndSetDownloadData(): Promise<boolean> {
        const result = await generateUacCodesForSampleFile(instrumentName, file);
        setDownloadData(result);
        return result.length > 0;
    }

    async function _handleSubmit() {
        switch (activeStep) {
            case 0:
                setActiveStep(await sampleFileAlreadyExists(instrumentName) ? 1 : 2);
                break;
            case 1:
                setActiveStep(overwrite === "Yes" ? 2 : 3);
                break;
            case 2:
                setActiveStep(await generateUacCodesAndSetDownloadData() ? 3 : 4);
                break;
            default:
                setActiveStep(0);
        }
    }

    return (
        <>
            <main id="main-content" className="page__main u-mt-no">
                <Switch>
                    <Route path="/app">
                        <App/>
                    </Route>
                    <Route path="/">

                        <ul className="list list--bare list--inline u-mt-m">
                            <li className="list__item">
                                <Link to="/app" id="generate-uac-link">
                                    Home
                                </Link>
                            </li>
                        </ul>
                    </Route>
                </Switch>

                <Formik
                    validateOnBlur={false}
                    validateOnChange={false}
                    initialValues={{override: ""}}
                    onSubmit={_handleSubmit}>

                    <Form id={"formID"}>
                        <div>
                            <div className="u-mt-m">
                                {_renderStepContent(activeStep)}
                            </div>
                            <div className="btn-group u-mt-m">
                                <ONSButton
                                    id={"next-screen-button"}
                                    submit={true}
                                    primary={true} label={"Continue"}/>
                            </div>
                        </div>
                    </Form>

                </Formik>
            </main>
        </>
    );
}

export default UploadSamplePage;
