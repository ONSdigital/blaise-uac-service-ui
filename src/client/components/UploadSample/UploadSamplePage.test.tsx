import React from "react";
import {Router} from "react-router";
import {render, waitFor, fireEvent, cleanup, screen} from "@testing-library/react";
import "@testing-library/jest-dom";
import {createMemoryHistory} from "history";
import UploadSamplePage from "./UploadSamplePage";


describe("Upload Sample Page", () => {

    it("select file page matches Snapshot", async () => {
        const history = createMemoryHistory();
        const wrapper = render(
            <Router history={history}>
                <UploadSamplePage/>
            </Router>
        );

        await waitFor(() => {
            expect(wrapper).toMatchSnapshot();
        });
    });

    it("should render correctly", async () => {
        const history = createMemoryHistory();
        const {queryByText} = render(
            <Router history={history}>
                <UploadSamplePage/>
            </Router>
        );

        expect(queryByText(/Select instrument sample/i)).toBeInTheDocument();
    });

    it("should display an error message if you dont enter an instrument name", async () => {
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <UploadSamplePage/>
            </Router>
        );

        await fireEvent.click(screen.getByText(/Continue/));

        await waitFor(() => {
            expect(screen.queryAllByText("Enter a valid instrument name")).toHaveLength(2);
        });
    });

    it("should display an error message if you dont select a file", async () => {
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <UploadSamplePage/>
            </Router>
        );

        await fireEvent.click(screen.getByText(/Continue/));

        await waitFor(() => {
            expect(screen.queryAllByText("Select a file")).toHaveLength(2);
        });
    });
});