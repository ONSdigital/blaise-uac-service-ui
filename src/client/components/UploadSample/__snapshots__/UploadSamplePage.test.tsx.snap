// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload Sample Page select file page matches Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="page__main u-mt-no"
        id="main-content"
      >
        <ul
          class="list list--bare list--inline u-mt-m"
        >
          <li
            class="list__item"
          >
            <a
              href="/app"
              id="deploy-questionnaire-link"
            >
              Home
            </a>
          </li>
        </ul>
        <form
          action="#"
          id="formID"
        >
          <div>
            <h1
              class="u-mt-s"
            >
              Select instrument sample
            </h1>
            <div
              class="panel panel--info panel--no-title  u-mt-m"
            >
              <div
                class="panel__body "
              >
                <p>
                  When a sample file is selected and you continue to upload this sample file, 
                  <b>
                    this may take a few minutes
                  </b>
                  .
                  <br />
                  <br />
                  Given this, 
                  <b>
                    do not navigate away
                  </b>
                   from this page during this process. You will be re-directed when there is an update regarding the deploy of the questionnaire.
                </p>
              </div>
            </div>
          </div>
          <div
            class="u-mt-m"
          >
            <div
              class="field"
            >
              <label
                class="label "
                for="instrument-name"
              >
                Instrument name
              </label>
              <input
                class="input input--text input-type__input"
                id="instrument-name"
                name="Instrument name"
                type="input"
                value=""
              />
            </div>
            <div
              class="field"
            >
              <label
                class="label label--with-description"
                for="sample-selector"
              >
                Select sample file
              </label>
              <span
                class="label__description  input--with-description"
                id="description-hint"
              >
                File type accepted is .csv
              </span>
              <input
                accept=".csv"
                class="input input--text input-type__input input--upload"
                id="sample-selector"
                name="Select sample file"
                type="file"
                value=""
              />
            </div>
          </div>
          <div
            class="btn-group u-mt-m"
          >
            <button
              class="btn "
              data-testid="button"
              id="upload-sample-button"
              type="submit"
            >
              <span
                class="btn__inner"
              >
                Continue
              </span>
            </button>
          </div>
        </form>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="page__main u-mt-no"
      id="main-content"
    >
      <ul
        class="list list--bare list--inline u-mt-m"
      >
        <li
          class="list__item"
        >
          <a
            href="/app"
            id="deploy-questionnaire-link"
          >
            Home
          </a>
        </li>
      </ul>
      <form
        action="#"
        id="formID"
      >
        <div>
          <h1
            class="u-mt-s"
          >
            Select instrument sample
          </h1>
          <div
            class="panel panel--info panel--no-title  u-mt-m"
          >
            <div
              class="panel__body "
            >
              <p>
                When a sample file is selected and you continue to upload this sample file, 
                <b>
                  this may take a few minutes
                </b>
                .
                <br />
                <br />
                Given this, 
                <b>
                  do not navigate away
                </b>
                 from this page during this process. You will be re-directed when there is an update regarding the deploy of the questionnaire.
              </p>
            </div>
          </div>
        </div>
        <div
          class="u-mt-m"
        >
          <div
            class="field"
          >
            <label
              class="label "
              for="instrument-name"
            >
              Instrument name
            </label>
            <input
              class="input input--text input-type__input"
              id="instrument-name"
              name="Instrument name"
              type="input"
              value=""
            />
          </div>
          <div
            class="field"
          >
            <label
              class="label label--with-description"
              for="sample-selector"
            >
              Select sample file
            </label>
            <span
              class="label__description  input--with-description"
              id="description-hint"
            >
              File type accepted is .csv
            </span>
            <input
              accept=".csv"
              class="input input--text input-type__input input--upload"
              id="sample-selector"
              name="Select sample file"
              type="file"
              value=""
            />
          </div>
        </div>
        <div
          class="btn-group u-mt-m"
        >
          <button
            class="btn "
            data-testid="button"
            id="upload-sample-button"
            type="submit"
          >
            <span
              class="btn__inner"
            >
              Continue
            </span>
          </button>
        </div>
      </form>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
